src/box/box.cc

```diff
@@ -2208,6 +2212,7 @@ box_make_bootstrap_leader(void)
 		return box_set_bootstrap_leader_record();
 	} else {
 		bootstrap_leader_uuid = INSTANCE_UUID;
+		is_supervised_bootstrap_leader = true;
```

Comment:
The same comment, as in message logging. AFAICS, we don't need that

Reply:
OK, removed and things seems working, but I'm a bit afraid that there is a point, when `bootstrap_leader_uuid` is not consistent with `is_supervised_bootstrap_leader`. That's how I described this contraint:

```c
/**
 * Whether the given instance is configured to make bootstrap of
 * the replicaset.
 *
 * Has an effect only in the supervised bootstrap strategy.
 *
 * This flag have to be consistent with bootstrap_leader_uuid
 * except the following cases:
 *
 * - INSTANCE_UUID is zero UUID
 * - the bootstrap strategy is not 'supervised'
 */
extern bool is_supervised_bootstrap_leader;
```

Reply:
> when bootstrap_leader_uuid is not consistent with is_supervised_bootstrap_leader

I was wrong. Not only the flag is not consistent with `bootstrap_leader_uuid`, but it's also looks like that it's not guaranteed, that we become a bootsrap leader, since `replicaset_find_join_master` may return another instance, not NULL or ourselves (especially, when applier to the current instance doesn't exist), `is_supervised_bootstrap_leader` will be `false` at that point.

The tests are stable, since this situation is not covered: we have only the test before the `box.cfg` but not during it (as it's done in `test_bootstrap`, but there only one `make_bootstrap_leader` is called). I won't insist on adding this test, since it seems it will be difficult to stabilize.

However, let's revert that change, flag should be set here. Sorry

Reply:
Thank you for another glance!

I've tried to create a test case and found that it actually depends on an exact point of time, when we call `box.ctl.make_bootstrap_leader()` via console.

It mostly fails before the update, but passes 3/4 times afterwards.

I'll not add it into the pull request (don't know how to stabilize), but here it is:

<details>
<summary><code>test_several_leaders</code></summary>

```lua
g_supervised.before_test('test_several_leaders', function(cg)
    cg.replica_set = replica_set:new{}
    cg.box_cfg_1 = {
        bootstrap_strategy = 'supervised',
        -- Has only itself in upstreams and can bootstrap on its
        -- own.
        replication = {
            server.build_listen_uri('server1', cg.replica_set.id),
        },
        replication_timeout = 0.1,
    }
    cg.box_cfg_2 = {
        bootstrap_strategy = 'supervised',
        -- Has no itself in upstreams, but has another instance
        -- that advertises itself as a bootstrap leader.
        --
        -- The expectation is that server2 anyway bootstraps its
        -- own database, because box.ctl.make_bootstrap_leader()
        -- is called before box.cfg().
        replication = {
            server.build_listen_uri('server1', cg.replica_set.id),
        },
        replication_timeout = 0.1,
    }
    for i = 1, 2 do
        local alias = 'server' .. i
        cg[alias] = cg.replica_set:build_and_add_server{
            alias = alias,
            box_cfg = cg['box_cfg_' .. i],
        }
        cg['admin' .. i] = fio.pathjoin(cg[alias].workdir, 'server.admin')
        local run_before_cfg = string.format([[
            local console = require('console')
            console.listen('unix/:%s')
        ]], cg['admin' .. i])
        cg[alias].env.TARANTOOL_RUN_BEFORE_BOX_CFG = run_before_cfg
    end
end)

-- XXX
g_supervised.test_several_leaders = function(cg)
    cg.server1:start({wait_until_ready = false})
    t.helpers.retrying({}, make_bootstrap_leader_initial, cg.admin1)

    cg.server2:start({wait_until_ready = false})
    t.helpers.retrying({}, make_bootstrap_leader_initial, cg.admin2)

    local query = 'ER_REPLICASET_UUID_MISMATCH'
    t.helpers.retrying({}, function()
        local found = grep_log(cg.server2, query)
        t.assert(found, ('Found %s'):format(query))
    end)
end
```
</details>

Reply:
Here's the stable version. Note, that in your version `box_cfg_1.replication` is incorrect.

<details><summary>Details</summary>
<p>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index f4761f283f..f15df94f42 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -5386,6 +5386,7 @@ bootstrap(bool *is_bootstrap_leader)
 	bool say_once = false;
 
 	while (true) {
+		ERROR_INJECT_YIELD(ERRINJ_BOOTSTRAP_DELAY);
 		box_restart_replication();
 		master = replicaset_find_join_master();
 		bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index b2b8ffa9e2..39ab38c10a 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -79,6 +79,7 @@ struct errinj {
 	_(ERRINJ_APPLIER_SLOW_ACK, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_APPLIER_STOP_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_APPLIER_SUBSCRIBE_DELAY, ERRINJ_BOOL, {.bparam = false}) \
+	_(ERRINJ_BOOTSTRAP_DELAY, ERRINJ_BOOL, {.bparam = true}) \
 	_(ERRINJ_BUILD_INDEX, ERRINJ_INT, {.iparam = -1}) \
 	_(ERRINJ_BUILD_INDEX_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_BUILD_INDEX_DISABLE_YIELD, ERRINJ_BOOL, {.bparam = false}) \
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index 1a510a42e0..a63f850101 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -539,18 +539,22 @@ local g_supervised = t.group('gh-8509-bootstrap-strategy-supervised')
 local server2_admin
 local SOCKET_TIMEOUT = 5
 
-local function make_bootstrap_leader_initial(sockname)
+local function make_socket_exec(sockname, command, resp)
     local sock, err = socket.tcp_connect('unix/', sockname)
     t.assert_equals(err, nil, 'Connection successful')
     local greeting = sock:read(128, SOCKET_TIMEOUT)
     t.assert_str_contains(greeting, 'Tarantool', 'Connected to console')
     t.assert_str_contains(greeting, 'Lua console', 'Connected to console')
-    sock:write('box.ctl.make_bootstrap_leader()\n', SOCKET_TIMEOUT)
+    sock:write(command, SOCKET_TIMEOUT)
     local response = sock:read(8, SOCKET_TIMEOUT)
-    t.assert_equals(response, '---\n...\n', 'The call succeeded')
+    t.assert_equals(response, resp, 'The call succeeded')
     sock:close()
 end
 
+local function make_bootstrap_leader_initial(sockname)
+    make_socket_exec(sockname, 'box.ctl.make_bootstrap_leader()\n', '---\n...\n')
+end
+
 g_supervised.after_each(function(cg)
     cg.replica_set:drop()
 end)
@@ -889,24 +893,34 @@ g_supervised.before_test('test_early_leader_several_leaders', function(cg)
     cg.box_cfg = {
         bootstrap_strategy = 'supervised',
         replication = {
-            server.build_listen_uri('server1', cg.replica_set.id),
             server.build_listen_uri('server2', cg.replica_set.id),
         },
         replication_timeout = 0.1,
     }
-    for i = 1, 2 do
-        local alias = 'server' .. i
-        cg[alias] = cg.replica_set:build_and_add_server{
-            alias = alias,
-            box_cfg = cg.box_cfg,
-            -- Both instances advertise itself as bootstrap
-            -- leaders.
-            env = {
-                ['TARANTOOL_RUN_BEFORE_BOX_CFG'] =
-                    'box.ctl.make_bootstrap_leader()',
-            },
-        }
-    end
+    cg['server1'] = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = cg.box_cfg,
+        -- Both instances advertise itself as bootstrap
+        -- leaders.
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] =
+                'box.ctl.make_bootstrap_leader()',
+        },
+    }
+
+    cg.box_cfg.replication = {
+        server.build_listen_uri('server1', cg.replica_set.id),
+    }
+    cg['server2'] = cg.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = cg.box_cfg,
+        -- Both instances advertise itself as bootstrap
+        -- leaders.
+        env = {
+            ['TARANTOOL_RUN_BEFORE_BOX_CFG'] =
+                'box.ctl.make_bootstrap_leader()',
+        },
+    }
 end)
 
 -- For the sake of completeness verify behavior in the incorrect
@@ -921,3 +935,63 @@ g_supervised.test_early_leader_several_leaders = function(cg)
         t.assert(found, ('Found %s'):format(query))
     end)
 end
+
+g_supervised.before_test('test_several_leaders', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.box_cfg_1 = {
+        bootstrap_strategy = 'supervised',
+        -- Has only itself in upstreams and can bootstrap on its
+        -- own.
+        replication = {
+            server.build_listen_uri('server2', cg.replica_set.id),
+        },
+        replication_timeout = 0.1,
+    }
+    cg.box_cfg_2 = {
+        bootstrap_strategy = 'supervised',
+        -- Has no itself in upstreams, but has another instance
+        -- that advertises itself as a bootstrap leader.
+        --
+        -- The expectation is that server2 anyway bootstraps its
+        -- own database, because box.ctl.make_bootstrap_leader()
+        -- is called before box.cfg().
+        replication = {
+            server.build_listen_uri('server1', cg.replica_set.id),
+        },
+        replication_timeout = 0.1,
+    }
+    for i = 1, 2 do
+        local alias = 'server' .. i
+        cg[alias] = cg.replica_set:build_and_add_server{
+            alias = alias,
+            box_cfg = cg['box_cfg_' .. i],
+        }
+        cg['admin' .. i] = fio.pathjoin(cg[alias].workdir, 'server.admin')
+        local run_before_cfg = string.format([[
+            box.error.injection.set('ERRINJ_BOOTSTRAP_DELAY', true)
+            local console = require('console')
+            console.listen('unix/:%s')
+        ]], cg['admin' .. i])
+        cg[alias].env.TARANTOOL_RUN_BEFORE_BOX_CFG = run_before_cfg
+    end
+end)
+
+-- XXX
+g_supervised.test_several_leaders = function(cg)
+    cg.server1:start({wait_until_ready = false})
+    t.helpers.retrying({}, make_bootstrap_leader_initial, cg.admin1)
+
+    cg.server2:start({wait_until_ready = false})
+    t.helpers.retrying({}, make_bootstrap_leader_initial, cg.admin2)
+
+    local cmd = "box.error.injection.set('ERRINJ_BOOTSTRAP_DELAY', false)\n"
+    local response = '---\n- ok'
+    t.helpers.retrying({}, make_socket_exec, cg.admin1, cmd, response)
+    t.helpers.retrying({}, make_socket_exec, cg.admin2, cmd, response)
+
+    local query = 'ER_REPLICASET_UUID_MISMATCH'
+    t.helpers.retrying({}, function()
+        local found = grep_log(cg.server2, query)
+        t.assert(found, ('Found %s'):format(query))
+    end)
+end

```

</p>
</details>